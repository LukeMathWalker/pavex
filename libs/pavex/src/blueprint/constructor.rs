use crate::blueprint::ErrorHandler;
use crate::blueprint::Lint;
use crate::blueprint::conversions::{
    cloning2cloning, coordinates2coordinates, lifecycle2lifecycle, lint2lint,
};
use pavex_bp_schema::Component;
use pavex_bp_schema::{Blueprint as BlueprintSchema, LintSetting, Location};

use super::CloningPolicy;
use super::Lifecycle;
use super::reflection::AnnotationCoordinates;

/// The input type for [`Blueprint::constructor`].
///
/// Check out [`Blueprint::constructor`] for more information on dependency injection
/// in Pavex.
///
/// # Stability guarantees
///
/// Use one of Pavex's constructor attributes (
/// [`singleton`](macro@crate::singleton), [`request_scoped`](macro@crate::request_scoped), or [`transient`](macro@crate::transient))
/// to create instances of `Constructor`.\
/// `Constructor`'s fields are an implementation detail of Pavex's macros and should not be relied upon:
/// newer versions of Pavex may add, remove or modify its fields.
///
/// [`Blueprint::constructor`]: crate::Blueprint::constructor
pub struct Constructor {
    #[doc(hidden)]
    pub coordinates: AnnotationCoordinates,
}

/// A constructor registered via [`Blueprint::constructor`].
///
/// # Example
///
/// You can use the methods exposed by [`RegisteredConstructor`] to tune the behaviour
/// of the registered constructor type.
/// For example, instruct Pavex to clone the constructed type if it's necessary to satisfy
/// the borrow checker:
///
/// ```rust
/// use pavex::{methods, Blueprint};
///
/// # pub struct PoolConfig;
/// pub struct Pool {
///     // [...]
/// }
///
/// #[methods]
/// impl Pool {
///     #[singleton]
///     pub fn new(config: &PoolConfig) -> Self {
///         # todo!()
///         // [...]
///     }
/// }
///
/// let mut bp = Blueprint::new();
/// // This is equivalent to `#[singleton(clone_if_necessary)]`
/// bp.constructor(POOL_NEW).clone_if_necessary();
/// ```
///
/// # Example: override the annotation
///
/// You can also override the behaviour specified via the [`singleton`](macro@crate::singleton) attribute.
///
/// ```rust
/// use pavex::{methods, Blueprint};
///
/// # pub struct PoolConfig;
/// pub struct Pool {
///     // [...]
/// }
///
/// #[methods]
/// impl Pool {
///     #[singleton(clone_if_necessary)]
///     pub fn new(config: &PoolConfig) -> Self {
///         # todo!()
///         // [...]
///     }
/// }
///
/// let mut bp = Blueprint::new();
/// // Using `never_clone` here, we are overriding the `clone_if_necessary`
/// // flag specified via the `singleton` attribute.
/// // This is equivalent to `#[singleton]`, thus restoring
/// // the default behaviour.
/// bp.constructor(POOL_NEW).never_clone();
/// ```
///
/// [`Blueprint::constructor`]: crate::Blueprint::constructor
pub struct RegisteredConstructor<'a> {
    pub(crate) blueprint: &'a mut BlueprintSchema,
    /// The index of the registered constructor in the blueprint's `components` vector.
    pub(crate) component_id: usize,
}

impl RegisteredConstructor<'_> {
    #[track_caller]
    /// Register an error handler.
    ///
    /// If an error handler has already been registered for this constructor, it will be
    /// overwritten.
    ///
    /// # Guide
    ///
    /// Check out the ["Error handlers"](https://pavex.dev/docs/guide/errors/error_handlers)
    /// section of Pavex's guide for a thorough introduction to error handlers
    /// in Pavex applications.
    ///
    /// # Example
    ///
    /// ```rust
    /// use pavex::Blueprint;
    /// use pavex::response::Response;
    /// use pavex::{methods, transient};
    /// # struct LogLevel;
    /// # struct Logger;
    /// # struct ConfigurationError;
    ///
    /// // ðŸ‘‡ a fallible constructor
    /// #[transient]
    /// pub fn logger() -> Result<Logger, ConfigurationError> {
    ///     // [...]
    ///     # todo!()
    /// }
    ///
    /// #[methods]
    /// impl ConfigurationError {
    ///     #[error_handler]
    ///     fn to_response(
    ///         #[px(error_ref)] &self,
    ///         log_level: LogLevel,
    ///     ) -> Response {
    ///         // [...]
    ///         # todo!()
    ///     }
    /// }
    ///
    /// # fn main() {
    /// let mut bp = Blueprint::new();
    /// bp.constructor(LOGGER)
    ///     .error_handler(CONFIGURATION_ERROR_TO_RESPONSE);
    /// # }
    /// ```
    pub fn error_handler(mut self, error_handler: ErrorHandler) -> Self {
        let error_handler = pavex_bp_schema::ErrorHandler {
            coordinates: coordinates2coordinates(error_handler.coordinates),
            registered_at: Location::caller(),
        };
        self.constructor().error_handler = Some(error_handler);
        self
    }

    /// Change the constructor lifecycle.
    pub fn lifecycle(mut self, lifecycle: Lifecycle) -> Self {
        self.constructor().lifecycle = Some(lifecycle2lifecycle(lifecycle));
        self
    }

    /// Set the cloning strategy for the output type returned by this constructor.
    ///
    /// By default,
    /// Pavex will **never** try to clone the output type returned by a constructor.
    /// If the output type implements [`Clone`], you can change the default by invoking
    /// [`clone_if_necessary`](Self::clone_if_necessary): Pavex will clone the output type if
    /// it's necessary to generate code that satisfies Rust's borrow checker.
    pub fn cloning(mut self, strategy: CloningPolicy) -> Self {
        self.constructor().cloning_policy = Some(cloning2cloning(strategy));
        self
    }

    /// Set the cloning strategy to [`CloningPolicy::CloneIfNecessary`].
    /// Check out [`cloning`](Self::cloning) method for more details.
    pub fn clone_if_necessary(self) -> Self {
        self.cloning(CloningPolicy::CloneIfNecessary)
    }

    /// Set the cloning strategy to [`CloningPolicy::NeverClone`].
    /// Check out [`cloning`](Self::cloning) method for more details.
    pub fn never_clone(self) -> Self {
        self.cloning(CloningPolicy::NeverClone)
    }

    /// Silence a specific [`Lint`] for this constructor.
    pub fn allow(mut self, lint: Lint) -> Self {
        self.constructor()
            .lints
            .insert(lint2lint(lint), LintSetting::Ignore);
        self
    }

    /// Fail the build if a specific [`Lint`] triggers
    /// for this constructor.
    pub fn deny(mut self, lint: Lint) -> Self {
        self.constructor()
            .lints
            .insert(lint2lint(lint), LintSetting::Enforce);
        self
    }

    fn constructor(&mut self) -> &mut pavex_bp_schema::Constructor {
        let component = &mut self.blueprint.components[self.component_id];
        let Component::Constructor(c) = component else {
            unreachable!("The component should be a constructor")
        };
        c
    }
}
