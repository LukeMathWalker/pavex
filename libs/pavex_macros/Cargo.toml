[package]
name = "pavex_macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = "2.0"

[dev-dependencies]
trybuild = "1.0"
pavex = { path = "../pavex" }
serde = { version = "1.0", features = ["derive"] }
