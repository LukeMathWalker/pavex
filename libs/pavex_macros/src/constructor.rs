use darling::FromMeta;
use proc_macro::TokenStream;
use quote::{ToTokens, quote};

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
pub struct Properties {
    pub lifecycle: Lifecycle,
    pub cloning_strategy: Option<CloningStrategy>,
    pub error_handler: Option<String>,
}

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
#[darling(rename_all = "snake_case")]
pub enum Lifecycle {
    Singleton,
    RequestScoped,
    Transient,
}

impl Lifecycle {
    pub fn as_str(&self) -> &'static str {
        match self {
            Lifecycle::Singleton => "singleton",
            Lifecycle::RequestScoped => "request_scoped",
            Lifecycle::Transient => "transient",
        }
    }
}

impl ToTokens for Lifecycle {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let s = self.as_str();
        tokens.extend(quote! { #s });
    }
}

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
#[darling(rename_all = "snake_case")]
pub enum CloningStrategy {
    CloneIfNecessary,
    NeverClone,
}
impl CloningStrategy {
    pub fn as_str(&self) -> &'static str {
        match self {
            CloningStrategy::CloneIfNecessary => "clone_if_necessary",
            CloningStrategy::NeverClone => "never_clone",
        }
    }
}

impl ToTokens for CloningStrategy {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let s = self.as_str();
        tokens.extend(quote! { #s });
    }
}

pub fn constructor(metadata: TokenStream, input: TokenStream) -> TokenStream {
    let attrs = match darling::ast::NestedMeta::parse_meta_list(metadata.into()) {
        Ok(attrs) => attrs,
        Err(err) => return err.to_compile_error().into(),
    };
    let Properties {
        lifecycle,
        cloning_strategy,
        error_handler,
    } = match Properties::from_list(&attrs) {
        Ok(parsed) => parsed,
        Err(err) => return err.write_errors().into(),
    };
    let mut properties = quote! {
        lifecycle = #lifecycle,
    };
    if let Some(cloning_strategy) = cloning_strategy {
        properties.extend(quote! {
            cloning_strategy = #cloning_strategy,
        });
    }
    if let Some(error_handler) = error_handler {
        properties.extend(quote! {
            error_handler = #error_handler,
        });
    }

    let input: proc_macro2::TokenStream = input.into();
    quote! {
        #[diagnostic::pavex::constructor(#properties)]
        #input
    }
    .into()
}
