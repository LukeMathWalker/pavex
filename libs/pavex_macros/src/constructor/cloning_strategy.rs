use quote::{ToTokens, quote};

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
#[darling(rename_all = "snake_case")]
pub enum CloningStrategy {
    CloneIfNecessary,
    NeverClone,
}
impl CloningStrategy {
    pub fn as_str(&self) -> &'static str {
        match self {
            CloningStrategy::CloneIfNecessary => "clone_if_necessary",
            CloningStrategy::NeverClone => "never_clone",
        }
    }
}

impl ToTokens for CloningStrategy {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let s = self.as_str();
        tokens.extend(quote! { #s });
    }
}
