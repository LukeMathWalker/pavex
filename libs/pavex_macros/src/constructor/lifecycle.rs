use quote::{ToTokens, quote};

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
#[darling(rename_all = "snake_case")]
pub enum Lifecycle {
    Singleton,
    RequestScoped,
    Transient,
}

impl Lifecycle {
    pub fn as_str(&self) -> &'static str {
        match self {
            Lifecycle::Singleton => "singleton",
            Lifecycle::RequestScoped => "request_scoped",
            Lifecycle::Transient => "transient",
        }
    }
}

impl ToTokens for Lifecycle {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let s = self.as_str();
        tokens.extend(quote! { #s });
    }
}
