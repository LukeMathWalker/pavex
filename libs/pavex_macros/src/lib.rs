use proc_macro::TokenStream;

use quote::quote;
use syn::punctuated::Punctuated;
use syn::{parse_macro_input, Attribute, Data, DeriveInput, Error, GenericParam, Token};

#[allow(non_snake_case)]
#[proc_macro_attribute]
pub fn RouteParams(_metadata: TokenStream, input: TokenStream) -> TokenStream {
    let ast = parse_macro_input!(input as DeriveInput);

    if let Err(mut e) = reject_serde_attributes(&ast) {
        // We emit both the error AND the original struct.
        // This is useful to avoid spurious "the type doesn't exist anymore" additional errors
        e.extend(vec![TokenStream::from(quote! { #ast })]);
        return e;
    }

    let generics_with_bounds = &ast.generics;
    let generics_without_bounds = &ast
        .generics
        .params
        .iter()
        .map(|p| match p {
            GenericParam::Lifetime(l) => {
                let l = &l.lifetime;
                quote! { #l }
            }
            GenericParam::Type(t) => {
                let i = &t.ident;
                quote! { #i }
            }
            GenericParam::Const(c) => {
                let i = &c.ident;
                quote! { #i }
            }
        })
        .collect::<Punctuated<_, Token![,]>>();
    let mut attrs = ast.attrs.clone();
    attrs.push(syn::parse_quote!(#[derive(serde::Serialize, serde::Deserialize)]));
    let struct_name = &ast.ident;
    let expanded = quote! {
        #(#attrs)*
        #ast

        impl #generics_with_bounds pavex_runtime::serialization::StructuralDeserialize for #struct_name < #generics_without_bounds > {}
    };

    TokenStream::from(expanded)
}

fn reject_serde_attributes(ast: &DeriveInput) -> Result<(), TokenStream> {
    for attr in &ast.attrs {
        reject_serde_attribute(attr)?;
    }
    match &ast.data {
        Data::Struct(data) => {
            for field in &data.fields {
                for attr in &field.attrs {
                    reject_serde_attribute(attr)?;
                }
            }
        }
        Data::Enum(data) => {
            for variant in &data.variants {
                for attr in &variant.attrs {
                    reject_serde_attribute(attr)?;
                }
                for field in &variant.fields {
                    for attr in &field.attrs {
                        reject_serde_attribute(attr)?;
                    }
                }
            }
        }
        Data::Union(union) => {
            for field in &union.fields.named {
                for attr in &field.attrs {
                    reject_serde_attribute(attr)?;
                }
            }
        }
    }
    Ok(())
}

/// We don't want to allow `serde` attributes on the top-level struct or any of its fields,
/// because we rely on `serde`'s default behaviour to determine, at code-generartion time,
/// if the route params can be deserialized from the URL of the incoming request.
fn reject_serde_attribute(attr: &Attribute) -> Result<(), TokenStream> {
    let err_msg = "`RouteParams` does not support `serde` attributes on the top-level struct or any of its fields.\n\n\
      `RouteParams` takes care of deriving `serde::Serialize` and `serde::Deserialize` for your struct, using the default \
       configuration. This allow Pavex to determine, at code-generation time, if the route params can \
       be successfully extracted from the URL of incoming requests for the relevant routes (e.g. do you \
       have a named field that doesn't map to any of the registered route parameters?).\n\n\
       If the default `serde` configuration won't work for your case, you should not derive `RouteParams` and \
       opt instead for implementing `serde::Serialize` and `serde::Deserialize` directly for your struct (either \
       manually or using a derive with custom attributes).\nKeep in mind that by going down this route \
       you give up compile-time checking of the route parameters!";
    if let Some(ident) = attr.path().get_ident() {
        if ident == "serde" {
            return Err(Error::new_spanned(attr, err_msg)
                .into_compile_error()
                .into());
        }
    }
    Ok(())
}
