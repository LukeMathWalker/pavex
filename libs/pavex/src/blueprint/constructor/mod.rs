//! Register constructors for the types that can be injected into your request and error handlers.  
//!
//! # Guide
//!
//! Check out the ["Dependency injection"](https://pavex.dev/docs/guide/dependency_injection)
//! section of Pavex's guide for a thorough introduction to dependency injection
//! in Pavex applications.
pub use lifecycle::Lifecycle;

use crate::blueprint::internals::RegisteredCallable;
use crate::blueprint::reflection::{RawCallable, RawCallableIdentifiers};
use crate::blueprint::Blueprint;

mod lifecycle;

/// The type returned by [`Blueprint::constructor`].
///
/// It allows you to further configure the behaviour of the registered constructor.
pub struct Constructor<'a> {
    pub(crate) blueprint: &'a mut Blueprint,
    /// The index of the registered constructor in the blueprint's `constructors` vector.
    pub(crate) constructor_id: usize,
}

impl<'a> Constructor<'a> {
    #[track_caller]
    /// Register an error handler.
    ///
    /// Error handlers convert the error type returned by your constructor into an HTTP response.
    ///
    /// Error handlers CANNOT consume the error type, they must take a reference to the
    /// error as input.  
    /// Error handlers can have additional input parameters alongside the error, as long as there
    /// are constructors registered for those parameter types.
    ///
    /// ```rust
    /// use pavex::f;
    /// use pavex::blueprint::{Blueprint, constructor::Lifecycle};
    /// use pavex::response::Response;
    /// # struct LogLevel;
    /// # struct Logger;
    /// # struct ConfigurationError;
    ///
    /// fn logger() -> Result<Logger, ConfigurationError> {
    ///     // [...]
    ///     # todo!()
    /// }
    ///
    /// fn error_to_response(error: &ConfigurationError, log_level: LogLevel) -> Response {
    ///     // [...]
    ///     # todo!()
    /// }
    ///
    /// # fn main() {
    /// let mut bp = Blueprint::new();
    /// bp.constructor(f!(crate::logger), Lifecycle::Transient)
    ///     .error_handler(f!(crate::error_to_response));
    /// # }
    /// ```
    ///
    /// If an error handler has already been registered for the same error type, it will be
    /// overwritten.
    ///
    /// ## Common Errors
    ///
    /// Pavex will fail to generate the runtime code for your application if you register
    /// an error handler for an infallible constructor (i.e. a constructor that doesn't return
    /// a `Result`).
    pub fn error_handler(self, error_handler: RawCallable) -> Self {
        let callable_identifiers = RawCallableIdentifiers::from_raw_callable(error_handler);
        let callable = RegisteredCallable {
            callable: callable_identifiers,
            location: std::panic::Location::caller().into(),
        };
        self.blueprint.constructors[self.constructor_id].error_handler = Some(callable);
        self
    }

    /// Set the cloning strategy for the output type returned by this constructor.
    ///
    /// By default,
    /// Pavex will **never** try to clone the output type returned by a constructor.  
    /// If the output type implements [`Clone`], you change the default by setting the cloning strategy
    /// to [`CloningStrategy::CloneIfNecessary`]: Pavex will clone the output type if
    /// it's necessary to generate code that satisfies Rust's borrow checker.
    pub fn cloning(self, strategy: CloningStrategy) -> Self {
        self.blueprint.constructors[self.constructor_id].cloning_strategy = Some(strategy);
        self
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[non_exhaustive]
/// Determines whether Pavex is allowed to clone the output type returned by a constructor.
///
/// Check out [`Constructor::cloning`] for more information.
pub enum CloningStrategy {
    /// Pavex will **never** try clone the output type returned by the constructor.
    NeverClone,
    /// Pavex will only clone the output type returned by this constructor if it's
    /// necessary to generate code that satisfies Rust's borrow checker.
    CloneIfNecessary,
}
