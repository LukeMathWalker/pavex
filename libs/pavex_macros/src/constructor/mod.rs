use crate::utils::fn_like::{Callable, CallableAnnotation, ImplContext};
use crate::utils::id::{callable_id_def, default_id};
use crate::utils::{AnnotationCodegen, CloningStrategy, CloningStrategyFlags};
use darling::util::Flag;
use lifecycle::Lifecycle;
use quote::quote;

mod lifecycle;

#[derive(darling::FromMeta, Debug, Clone)]
/// The available options for `#[pavex::constructor]`.
pub struct InputSchema {
    pub id: Option<syn::Ident>,
    pub pavex: Option<syn::Ident>,
    pub singleton: Flag,
    pub request_scoped: Flag,
    pub transient: Flag,
    pub clone_if_necessary: Flag,
    pub never_clone: Flag,
    pub allow: Option<ConstructorAllows>,
}

#[derive(darling::FromMeta, Debug, Clone)]
pub struct ConstructorAllows {
    unused: Flag,
}

#[derive(darling::FromMeta, Debug, Clone)]
/// The available options for `#[pavex::request_scoped]`, `#[pavex::transient]`
/// and `#[pavex::singleton]`.
/// Everything in [`InputSchema`], minus `lifecycle`.
pub struct ShorthandSchema {
    pub id: Option<syn::Ident>,
    pub pavex: Option<syn::Ident>,
    pub clone_if_necessary: Flag,
    pub never_clone: Flag,
    pub allow: Option<ConstructorAllows>,
}

impl TryFrom<InputSchema> for Properties {
    type Error = darling::Error;

    fn try_from(input: InputSchema) -> Result<Self, Self::Error> {
        let InputSchema {
            singleton,
            request_scoped,
            transient,
            clone_if_necessary,
            never_clone,
            id,
            pavex,
            allow,
        } = input;

        let lifecycle = match (
            singleton.is_present(),
            request_scoped.is_present(),
            transient.is_present(),
        ) {
            (true, false, false) => Lifecycle::Singleton,
            (false, true, false) => Lifecycle::RequestScoped,
            (false, false, true) => Lifecycle::Transient,
            (false, false, false) => {
                return Err(darling::Error::custom(
                    "You must specify the lifecycle of your constructor. It can either be `singleton`, `request_scoped`, or `transient`",
                ));
            }
            _ => {
                return Err(darling::Error::custom(
                    "A constructor can't have multiple lifecycles. You can only specify *one* of `singleton`, `request_scoped`, or `transient`.",
                ));
            }
        };

        let Ok(cloning_strategy) = CloningStrategyFlags {
            clone_if_necessary,
            never_clone,
        }
        .try_into() else {
            return Err(darling::Error::custom(
                "A constructor can't have multiple cloning strategies. You can only specify *one* of `never_clone` and `clone_if_necessary`.",
            ));
        };

        let allow_unused = allow.as_ref().map(|a| a.unused.is_present());

        Ok(Properties {
            lifecycle,
            cloning_strategy,
            id,
            pavex,
            allow_unused,
        })
    }
}

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
pub struct Properties {
    pub lifecycle: Lifecycle,
    pub cloning_strategy: Option<CloningStrategy>,
    pub id: Option<syn::Ident>,
    pub pavex: Option<syn::Ident>,
    pub allow_unused: Option<bool>,
}

#[derive(darling::FromMeta, Debug, Clone, PartialEq, Eq)]
/// Everything in [`Properties`], minus `lifecycle`.
pub struct ShorthandProperties {
    pub cloning_strategy: Option<CloningStrategy>,
    pub id: Option<syn::Ident>,
    pub pavex: Option<syn::Ident>,
    pub allow_unused: Option<bool>,
}

impl TryFrom<ShorthandSchema> for ShorthandProperties {
    type Error = darling::Error;

    fn try_from(input: ShorthandSchema) -> Result<Self, Self::Error> {
        let ShorthandSchema {
            clone_if_necessary,
            never_clone,
            id,
            pavex,
            allow,
        } = input;
        let Ok(cloning_strategy) = CloningStrategyFlags {
            clone_if_necessary,
            never_clone,
        }
        .try_into() else {
            return Err(darling::Error::custom(
                "A constructor can't have multiple cloning strategies. You can only specify *one* of `never_clone` and `clone_if_necessary`.",
            ));
        };
        let allow_unused = allow.as_ref().map(|a| a.unused.is_present());

        Ok(Self {
            cloning_strategy,
            id,
            pavex,
            allow_unused,
        })
    }
}

pub struct ConstructorAnnotataion;

impl CallableAnnotation for ConstructorAnnotataion {
    const PLURAL_COMPONENT_NAME: &str = "Constructors";

    const ATTRIBUTE: &str = "#[pavex::constructor]";

    type InputSchema = InputSchema;

    fn codegen(
        impl_: Option<ImplContext>,
        metadata: Self::InputSchema,
        item: Callable,
    ) -> Result<AnnotationCodegen, proc_macro::TokenStream> {
        let properties = metadata
            .try_into()
            .map_err(|e: darling::Error| e.write_errors())?;
        Ok(emit(impl_, item, properties))
    }
}

pub struct RequestScopedAnnotation;

impl CallableAnnotation for RequestScopedAnnotation {
    const PLURAL_COMPONENT_NAME: &str = "Request-scoped constructors";

    const ATTRIBUTE: &str = "#[pavex::request_scoped]";

    type InputSchema = ShorthandSchema;

    fn codegen(
        impl_: Option<ImplContext>,
        metadata: Self::InputSchema,
        item: Callable,
    ) -> Result<AnnotationCodegen, proc_macro::TokenStream> {
        shorthand(impl_, metadata, item, Lifecycle::RequestScoped)
    }
}

pub struct TransientAnnotation;

impl CallableAnnotation for TransientAnnotation {
    const PLURAL_COMPONENT_NAME: &str = "Transient constructors";

    const ATTRIBUTE: &str = "#[pavex::transient]";

    type InputSchema = ShorthandSchema;

    fn codegen(
        impl_: Option<ImplContext>,
        metadata: Self::InputSchema,
        item: Callable,
    ) -> Result<AnnotationCodegen, proc_macro::TokenStream> {
        shorthand(impl_, metadata, item, Lifecycle::Transient)
    }
}

pub struct SingletonAnnotation;

impl CallableAnnotation for SingletonAnnotation {
    const PLURAL_COMPONENT_NAME: &str = "Singleton constructors";

    const ATTRIBUTE: &str = "#[pavex::singleton]";

    type InputSchema = ShorthandSchema;

    fn codegen(
        impl_: Option<ImplContext>,
        metadata: Self::InputSchema,
        item: Callable,
    ) -> Result<AnnotationCodegen, proc_macro::TokenStream> {
        shorthand(impl_, metadata, item, Lifecycle::Singleton)
    }
}

fn shorthand(
    impl_: Option<ImplContext>,
    schema: ShorthandSchema,
    item: Callable,
    lifecycle: Lifecycle,
) -> Result<AnnotationCodegen, proc_macro::TokenStream> {
    let ShorthandProperties {
        cloning_strategy,
        id,
        pavex,
        allow_unused,
    } = schema
        .try_into()
        .map_err(|e: darling::Error| e.write_errors())?;
    let properties = Properties {
        lifecycle,
        cloning_strategy,
        id,
        pavex,
        allow_unused,
    };
    Ok(emit(impl_, item, properties))
}

/// Decorate the input with a `#[diagnostic::pavex::constructor]` attribute
/// that matches the provided properties.
fn emit(impl_: Option<ImplContext>, item: Callable, properties: Properties) -> AnnotationCodegen {
    let Properties {
        lifecycle,
        cloning_strategy,
        id,
        pavex,
        allow_unused,
    } = properties;

    let id = id.unwrap_or_else(|| default_id(impl_.as_ref(), &item));
    let id_str = id.to_string();

    let mut properties = quote! {
        id = #id_str,
        lifecycle = #lifecycle,
    };
    if let Some(cloning_strategy) = cloning_strategy {
        properties.extend(quote! {
            cloning_strategy = #cloning_strategy,
        });
    }
    if let Some(allow_unused) = allow_unused {
        properties.extend(quote! {
            allow_unused = #allow_unused,
        });
    }

    AnnotationCodegen {
        id_def: Some(callable_id_def(
            &id,
            pavex.as_ref(),
            "constructor",
            "Constructor",
            "a constructor",
            "constructor",
            true,
            impl_.as_ref(),
            &item,
        )),
        new_attributes: vec![syn::parse_quote! {
            #[diagnostic::pavex::constructor(#properties)]
        }],
    }
}
