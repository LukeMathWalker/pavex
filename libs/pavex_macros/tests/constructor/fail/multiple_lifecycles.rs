use pavex_macros::constructor;

pub struct A;

#[constructor(singleton, request_scoped)]
pub fn new1() -> A {
    todo!()
}

#[constructor(singleton, transient)]
pub fn new2() -> A {
    todo!()
}

#[constructor(singleton, request_scoped, transient)]
pub fn new3() -> A {
    todo!()
}

#[constructor(request_scoped, transient)]
pub fn new4() -> A {
    todo!()
}

fn main() {}
