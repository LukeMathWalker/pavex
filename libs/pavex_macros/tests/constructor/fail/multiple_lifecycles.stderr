error: A constructor can't have multiple lifecycles. You can only specify *one* of `singleton`, `request_scoped`, or `transient`.
 --> tests/constructor/fail/multiple_lifecycles.rs:5:1
  |
5 | #[constructor(singleton, request_scoped)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `constructor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: A constructor can't have multiple lifecycles. You can only specify *one* of `singleton`, `request_scoped`, or `transient`.
  --> tests/constructor/fail/multiple_lifecycles.rs:10:1
   |
10 | #[constructor(singleton, transient)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `constructor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: A constructor can't have multiple lifecycles. You can only specify *one* of `singleton`, `request_scoped`, or `transient`.
  --> tests/constructor/fail/multiple_lifecycles.rs:15:1
   |
15 | #[constructor(singleton, request_scoped, transient)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `constructor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: A constructor can't have multiple lifecycles. You can only specify *one* of `singleton`, `request_scoped`, or `transient`.
  --> tests/constructor/fail/multiple_lifecycles.rs:20:1
   |
20 | #[constructor(request_scoped, transient)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `constructor` (in Nightly builds, run with -Z macro-backtrace for more info)
